src/microwasm.rs

```diff
@@ -0,0 +1,1696 @@
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Value {
+    I32(i32),
+    I64(i64),
+    F32(Ieee32),
+    F64(Ieee64),
+}
+
+impl fmt::Display for Value {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Value::I32(v) => write!(f, "{}i32", v),
+            Value::I64(v) => write!(f, "{}i64", v),
+            Value::F32(v) => write!(f, "{}f32", f32::from_bits(v.bits())),
+            Value::F64(v) => write!(f, "{}f64", f64::from_bits(v.bits())),
+        }
+    }
+}
+
+impl Value {
+    fn default_for_type(ty: SignlessType) -> Self {
+        match ty {
+            Type::Int(Size::_32) => Value::I32(0),
+            Type::Int(Size::_64) => Value::I64(0),
+            Type::Float(Size::_32) => Value::F32(Ieee32(0)),
+            Type::Float(Size::_64) => Value::F64(Ieee64(0)),
+        }
+    }
+}
+
+/// Whether to interpret an integer as signed or unsigned
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Signedness {
+    Signed,
+    Unsigned,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Size {
+    _32,
+    _64,
+}
+
+type Int = Size;
+type Float = Size;
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct SignfulInt(Signedness, Size);
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Type<I> {
+    Int(I),
+    Float(Size),
+}
+
+impl fmt::Display for SignfulType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Type::Int(i) => write!(f, "{}", i),
+            Type::Float(Size::_32) => write!(f, "f32"),
+            Type::Float(Size::_64) => write!(f, "f64"),
+        }
+    }
+}
+
+impl fmt::Display for SignlessType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Type::Int(Size::_32) => write!(f, "i32"),
+            Type::Int(Size::_64) => write!(f, "i64"),
+            Type::Float(Size::_32) => write!(f, "f32"),
+            Type::Float(Size::_64) => write!(f, "f64"),
+        }
+    }
+}
+
+impl fmt::Display for SignfulInt {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            SignfulInt(Signedness::Signed, Size::_32) => write!(f, "i32"),
+            SignfulInt(Signedness::Unsigned, Size::_32) => write!(f, "u32"),
+            SignfulInt(Signedness::Signed, Size::_64) => write!(f, "i64"),
+            SignfulInt(Signedness::Unsigned, Size::_64) => write!(f, "u64"),
+        }
+    }
+}
+
+pub type SignlessType = Type<Size>;
+pub type SignfulType = Type<SignfulInt>;
+
+pub const I32: SignlessType = Type::Int(Size::_32);
+pub const I64: SignlessType = Type::Int(Size::_64);
+pub const F32: SignlessType = Type::Float(Size::_32);
+pub const F64: SignlessType = Type::Float(Size::_64);
+
+pub mod sint {
+    use super::{Signedness, SignfulInt, Size};
+
+    pub const I32: SignfulInt = SignfulInt(Signedness::Signed, Size::_32);
+    pub const I64: SignfulInt = SignfulInt(Signedness::Signed, Size::_64);
```

Comment:
This is a good point. Naming these constants is actually quite difficult since we need to differentiate between a type that _could_ be signed but isn't and a type for which the sign is irrelevant